usr/libexec/helper-scripts/accountctl.sh

```diff
@@ -0,0 +1,395 @@
+    user_escaped="$(escape_name "${user}")"
+    pass="$(grep -e "^${user_escaped}:" -- /etc/shadow)"
+    pass="${pass#"${user}:"*}"
+    pass="${pass%%":"*}"
+  fi
+  printf '%s' "${pass}"
+}
+
+
+## Description: Get password without its state.
+## Output: Password prefix is trimmed.
+## Usage: get_clean_pass USER PREFIX
+## Example: get_clean_pass user '!*'
+get_clean_pass(){
+  log info "${FUNCNAME[0]} $*"
+  local user pass symbol
+  user="${1:-}"
+  is_user "${user}"
+  symbol="${2:-"!*"}"
+  pass="$(get_pass "${user}")"
+  local idx
+  for (( idx=0; idx < ${#symbol}; idx++ )); do
+    pass="${pass#"${symbol:${idx}:1}"}"
+  done
+  printf '%s\n' "${pass}"
+}
+
+
+## Description: Check if password is empty.
+## Output: None
+## Return: 0 when empty, 1 otherwise.
+## Usage: is_pass_empty USER
+## Example: is_pass_empty user
+is_pass_empty(){
+  log info "${FUNCNAME[0]} $*"
+  local user trim_pass
+  user="${1:-}"
+  is_user "${user}"
+  trim_pass="$(get_clean_pass "${user}" '!*')"
+  if test -z "${trim_pass}"; then
+    return 0
+  fi
+  return 1
+}
+
+
+## Description: Check if password is locked.
+## Output: None
+## Return: 0 when locked, 1 otherwise.
+## Usage: is_pass_locked USER
+## Example: is_pass_locked user
+is_pass_locked(){
+  log info "${FUNCNAME[0]} $*"
+  local user
+  user="${1:-}"
+  is_user "${user}"
+  case "$(get_pass "${user}")" in
+    "!"*) return 0;;
+    *) return 1;;
+  esac
+}
+
+
+## Description: Check if password is disabled.
+## Output: None
+## Return: 0 when disabled, 1 otherwise.
+## Usage: is_pass_disabled USER
+## Example: is_pass_disabled user
+is_pass_disabled(){
+  log info "${FUNCNAME[0]} $*"
+  local user
+  user="${1:-}"
+  is_user "${user}"
+  case "$(get_pass "${user}")" in
+    "*"*|"!*"*) return 0;;
+    *) return 1;;
+  esac
+}
+
+
+## Description: Lock a password.
+## Output: None
+## Return: passwd
+## Usage: lock_pass USER
+## Example: lock_pass user
+lock_pass(){
+  log info "${FUNCNAME[0]} $*"
+  has passwd
+  local user
+  user="${1:-}"
+  is_user "${user}"
+  if is_pass_locked "${user}"; then
+    return 0
+  fi
+  passwd -ql -- "${user}"
+}
+
+
+## Description: Unlock a password.
+## Output: None
+## Return: chpasswd
+## Usage: unlock_pass USER
+## Example: unlock_pass user
+unlock_pass(){
+  log info "${FUNCNAME[0]} $*"
+  has chpasswd
+  local user
+  user="${1:-}"
+  is_user "${user}"
+  if ! is_pass_locked "${user}"; then
+    return 0
+  fi
+  ## The tools "passwd" and "usermod" can't unlock when password is empty.
+  pass="$(get_clean_pass "${user}" '!')"
+  chpasswd -c NONE <<< "${user}:${pass}"
+}
+
+
+## Description: Disable a password.
+## Output: None
+## Return: chpasswd
+## Usage: disable_pass USER
+## Example: disable_pass user
+disable_pass(){
+  log info "${FUNCNAME[0]} $*"
+  has chpasswd
+  local user
+  user="${1:-}"
+  is_user "${user}"
+  if is_pass_disabled "${user}"; then
+    return 0
+  fi
+  pass="$(get_pass "${user}")"
+  chpasswd -c NONE <<< "${user}:*${pass}"
```

Comment:
This does not work to disable an account that is locked but not already disabled - the password will be set to `*!`, not `!*`, which probably has a different meaning to some tools. There might be a more efficient way to work around this, but this works:

```suggestion
  pass="$(get_clean_pass "${user}")"
  if is_pass_locked "${user}"; then
    chpasswd -c NONE <<< "${user}:!*${pass}"
    return 0
  fi
  chpasswd -c NONE <<< "${user}:*${pass}"
```